---
title: "Cell markers, TP53"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

Biallelic TP53
Biallelic status for gene (0 = bi-allelic inactive,  1 = one allele impacted, 2 = normal)
```{r BI_TP53}
genes20.mmrf <- readRDS(file="~/Downloads/Rshiny_cell-markers/Rdata/genes21/genes24_mmrf.rds")
p53 <- annotation$BI_TP53
names(p53) <- annotation$Patient
table(p53)
 #  0   1   2 
 # 21  41 483 
sum(is.na(p53)) # 53 NA's
### remove NA's
## check if patient order is same
all.equal(colnames(genes20.mmrf), names(p53)) # TRUE
genes20.mmrf <- genes20.mmrf[, !is.na(p53)] # 24 x 545
p53 <- p53[!is.na(p53)]
length(p53) # 545
table(p53)
p53 <- factor(p53)
levels(p53) ## This will show the levels of x
## To reorder the levels:
## note, if x is not a factor use levels(factor(x))
p53 <- factor(p53, levels(p53)[c(3, 2, 1)])
levels(p53)
head(p53)
saveRDS(p53, file="~/Downloads/Rshiny_cell-markers/Rdata/factors/BI_TP53_factors.rds")

### format dataset
## genes20.iss is only risk 1 vs 3
d <- data.frame(Genes=rownames(genes20.mmrf), genes20.mmrf)
dd <- d %>% gather(TP53, Expression, -Genes)
head(dd)
## change Patient to R-ISS
dd$TP53[which(dd$TP53 %in% names(p53)[p53==0])] <- 0
dd$TP53[which(dd$TP53 %in% names(p53)[p53==1])] <- 1
dd$TP53[which(dd$TP53 %in% names(p53)[p53==2])] <- 2
dd$TP53 <- factor(dd$TP53)
## To reorder the levels:
## note, if x is not a factor use levels(factor(x))
dd$TP53 <- factor(dd$TP53, levels(dd$TP53)[c(3, 2, 1)])
levels(dd$TP53)
table(dd$TP53)

### Save plot
tiff("~/Downloads/Rshiny_cell-markers/Rplots/figure1/BI_TP53_boxplot.tiff", res = 300, width = 10, height = 7, units = "in")
## specify how many groups need color
ncols <- 3
mycolors <- colorRampPalette(brewer.pal(9, "Blues"))(ncols)
## provide gene order from make_boxplots
genes.order <- readRDS("~/Downloads/Rshiny_cell-markers/Rdata/genes21/genes24_order.rds")
dd$Genes.relevel <- factor(dd$Genes, levels=genes.order)
## plot
ggplot(dd, aes(x=Genes.relevel, y=Expression, fill=TP53)) +
  geom_boxplot(notch=TRUE, outlier.shape=NA, show.legend=TRUE) +
  scale_fill_manual(values=mycolors) +
  theme(axis.text.x = element_text(angle=90, size=10)) +
  theme(axis.title = element_text(size=12)) +
  ggtitle("Gene expression stratified by TP53 allele status") + 
  ylab("log2(Expression)") +
  xlab("Genes")
dev.off()

### ANOVA
## make sure the following is true before running
all.equal(colnames(genes20.mmrf), names(p53)) # TRUE
markers <- rownames(genes20.mmrf)
for(gene in markers) {
  gene.expr <- genes20.mmrf[gene, ]
  
  # ANOVA
  model <- aov(gene.expr ~ p53)
  aov.pval <- summary(aov(model))[[1]][["Pr(>F)"]][[1]]
  
  # if significant, then Tukey
  if(aov.pval <= 0.05) {
    print(paste("yes", gene))
    
    tukey <- TukeyHSD(model)
    
    write.csv(tukey$p53, file=paste0("~/Downloads/Rshiny_cell-markers/Rtables/BI_TP53_tukey/", gene, ".csv"))
  }
}
```


p53 mutations
```{r}
genes20.mmrf <- readRDS(file="~/Downloads/Rshiny_cell-markers/Rdata/genes21/genes24_mmrf.rds")
p53 <- annotation$NS_TP53
names(p53) <- annotation$Patient
#   0   1 
# 563  35
sum(is.na(p53)) # 0 NAs
table(p53)
length(p53) # 598
p53 <- factor(p53)
levels(p53) ## This will show the levels of x
## To reorder the levels:
## note, if x is not a factor use levels(factor(x))
p53 <- factor(p53, labels=c("No mut", "Mut"))
levels(p53)
p53 <- factor(p53, levels(p53)[c(2, 1)]) # re-level
table(p53)
saveRDS(p53, file="~/Downloads/Rshiny_cell-markers/Rdata/factors/NS_TP53_factors.rds")

### format dataset
## genes20.iss is only risk 1 vs 3
d <- data.frame(Genes=rownames(genes20.mmrf), genes20.mmrf)
dd <- d %>% gather(TP53, Expression, -Genes)
head(dd)
## change Patient to R-ISS
dd$TP53[which(dd$TP53 %in% names(p53)[p53=="Mut"])] <- "Mut"
dd$TP53[which(dd$TP53 %in% names(p53)[p53=="No mut"])] <- "No mut"
dd$TP53 <- factor(dd$TP53)
levels(dd$TP53)
## To reorder the levels:
## note, if x is not a factor use levels(factor(x))
dd$TP53 <- factor(dd$TP53, levels(dd$TP53)[c(2, 1)])
levels(dd$TP53)
table(dd$TP53)

### Save plot
tiff("~/Downloads/Rshiny_cell-markers/Rplots/figure1/NS_TP53_boxplot.tiff", res = 300, width = 10, height = 7, units = "in")
## specify how many groups need color
ncols <- 2
mycolors <- colorRampPalette(brewer.pal(5, "Blues"))(ncols)
## provide gene order from make_boxplots
genes.order <- readRDS("~/Downloads/Rshiny_cell-markers/Rdata/genes21/genes24_order.rds")
dd$Genes.relevel <- factor(dd$Genes, levels=genes.order)
## plot
ggplot(dd, aes(x=Genes.relevel, y=Expression, fill=TP53)) +
  geom_boxplot(notch=TRUE, outlier.shape=NA, show.legend=TRUE) +
  scale_fill_manual(values=mycolors) +
  theme(axis.text.x = element_text(angle=90, size=10)) +
  theme(axis.title = element_text(size=12)) +
  ggtitle("Gene expression stratified by TP53 mutation status") + 
  ylab("log2(Expression)") +
  xlab("Genes")
dev.off()

### ANOVA
## make sure the following is true before running
all.equal(colnames(genes20.mmrf), names(p53)) # TRUE
markers <- rownames(genes20.mmrf)
for(gene in markers) {
  gene.expr <- genes20.mmrf[gene, ]
  
  # ANOVA
  model <- aov(gene.expr ~ p53)
  aov.pval <- summary(aov(model))[[1]][["Pr(>F)"]][[1]]
  
  # if significant, then Tukey
  if(aov.pval <= 0.05) {
    print(paste("yes", gene))
    
    tukey <- TukeyHSD(model)
    
    write.csv(tukey$p53, file=paste0("~/Downloads/Rshiny_cell-markers/Rtables/NS_TP53_tukey/", gene, ".csv"))
  }
}
```


p53 copy number variation
```{r}
genes20.mmrf <- readRDS(file="~/Downloads/Rshiny_cell-markers/Rdata/genes21/genes24_mmrf.rds")
p53 <- annotation$TP53_Loss
names(p53) <- annotation$Patient
sum(is.na(p53)) # 53 NAs
### remove NA's
## check if patient order is same
all.equal(colnames(genes20.mmrf), names(p53)) # TRUE
genes20.mmrf <- genes20.mmrf[, !is.na(p53)] # 24 x 545
p53 <- p53[!is.na(p53)]
length(p53) # 545
table(p53)
 # -2   1   2   3   4 
 # 12  38 464  28   3 
### combine -2 with 1, combine 4 with 3
p53[p53==-2] <- 1
p53[p53==4] <- 3

## To reorder the levels:
## note, if x is not a factor use levels(factor(x))
p53 <- factor(p53)
levels(p53)
## re-name factors
levels(p53)[levels(p53)==1] <- "Loss"
levels(p53)[levels(p53)==2] <- "Normal"
levels(p53)[levels(p53)==3] <- "Amplification"
## re-level, with normal first
p53 <- factor(p53, levels(p53)[c(2, 1, 3)])
levels(p53)
table(p53)
head(p53)
saveRDS(p53, file="~/Downloads/Rshiny_cell-markers/Rdata/factors/CNV_TP53_factors.rds")

### format dataset
## genes20.iss is only risk 1 vs 3
d <- data.frame(Genes=rownames(genes20.mmrf), genes20.mmrf)
dd <- d %>% gather(TP53, Expression, -Genes)
head(dd)
## change Patient to R-ISS
dd$TP53[which(dd$TP53 %in% names(p53)[p53=="Loss"])] <- "Loss"
dd$TP53[which(dd$TP53 %in% names(p53)[p53=="Normal"])] <- "Normal"
dd$TP53[which(dd$TP53 %in% names(p53)[p53=="Amplification"])] <- "Amplification"
dd$TP53 <- factor(dd$TP53)
levels(dd$TP53)
## To reorder the levels:
## note, if x is not a factor use levels(factor(x))
dd$TP53 <- factor(dd$TP53, levels(dd$TP53)[c(3, 2, 1)])
levels(dd$TP53)
table(dd$TP53)

### Save plot
tiff("~/Downloads/Rshiny_cell-markers/Rplots/figure1/CNV_TP53_boxplot.tiff", res = 300, width = 10, height = 7, units = "in")
## specify how many groups need color
ncols <- 3
mycolors <- colorRampPalette(brewer.pal(5, "Blues"))(ncols)
## provide gene order from make_boxplots
genes.order <- readRDS("~/Downloads/Rshiny_cell-markers/Rdata/genes21/genes24_order.rds")
dd$Genes.relevel <- factor(dd$Genes, levels=genes.order)
## plot
ggplot(dd, aes(x=Genes.relevel, y=Expression, fill=TP53)) +
  geom_boxplot(notch=TRUE, outlier.shape=NA, show.legend=TRUE) +
  scale_fill_manual(values=mycolors) +
  theme(axis.text.x = element_text(angle=90, size=10)) +
  theme(axis.title = element_text(size=12)) +
  ggtitle("Gene expression stratified by TP53 copy number variance") + 
  ylab("log2(Expression)") +
  xlab("Genes")
dev.off()

### ANOVA
## make sure the following is true before running
all.equal(colnames(genes20.mmrf), names(p53)) # TRUE
markers <- rownames(genes20.mmrf)
for(gene in markers) {
  gene.expr <- genes20.mmrf[gene, ]
  
  # ANOVA
  model <- aov(gene.expr ~ p53)
  aov.pval <- summary(aov(model))[[1]][["Pr(>F)"]][[1]]
  
  # if significant, then Tukey
  if(aov.pval <= 0.05) {
    print(paste("yes", gene))
    
    tukey <- TukeyHSD(model)
    
    write.csv(tukey$p53, file=paste0("~/Downloads/Rshiny_cell-markers/Rtables/CNV_TP53_tukey/", gene, ".csv"))
  }
}

## t-test for sanity check on Tukey values
t.test(genes20.mmrf["CD200", which(p53=="Normal")], genes20.mmrf["CD200", which(p53=="Loss")])
```